### PR TITLE
Implement K8sBindingInjector.Uninject

### PR DESCRIPTION
**Pull request by arschles**
Fixes https://github.com/kubernetes-incubator/service-catalog/issues/169

### CODE REVIEW

File: pkg/controller/injector/k8s_binding_injector.go [hunk 1]
```diff
@@ -69,18 +68,24 @@ func (b *k8sBindingInjector) Inject(binding *servicecatalog.Binding, cred *broke
 }
 
 func (b *k8sBindingInjector) Uninject(binding *servicecatalog.Binding) error {
-	return fmt.Errorf("Not implemented")
+	// TODO: don't hardcode this namespace. https://github.com/kubernetes-incubator/service-catalog/issues/162
```
Comment (MHBauer) [inline pkg/controller/injector/k8s_binding_injector.go]: would the namespace come from the binding, or would it be a new argument?

Comment (MHBauer) [inline pkg/controller/injector/k8s_binding_injector.go]: dependent response being, if it's a new argument, add it, even if it's unused.

I think we would pull it out of the ObjectMeta in the binding.

No action necessary.

**Comment by MHBauer** [comment]
LGTM

File: pkg/controller/injector/k8s_binding_injector.go (hunk 1 above)
Comment (arschles) [inline pkg/controller/injector/k8s_binding_injector.go]: >I think we would pull it out of the ObjectMeta in the binding.

I'm with ya. :+1:

Comment (arschles) [inline pkg/controller/injector/k8s_binding_injector.go]: That's also consistent with the overall plan for bindings we decided upon at the previous face-to-face

Comment (arschles) [inline pkg/controller/injector/k8s_binding_injector.go]: Also check out https://github.com/kubernetes-incubator/service-catalog/issues/162 @MHBauer

File: pkg/controller/injector/k8s_binding_injector.go [hunk 2]
```diff
@@ -60,7 +60,6 @@ func CreateK8sBindingInjector() (BindingInjector, error) {
 
 func (b *k8sBindingInjector) Inject(binding *servicecatalog.Binding, cred *brokerapi.Credential) error {
 	is := makeInjectionSet(binding, cred)
-
 	if err := b.injectSecret(is.secret); err != nil {
```
Comment (duglin) [inline pkg/controller/injector/k8s_binding_injector.go]: this may just be a stylist thing, I'm not sure, but the level of indirection here confuses me.
We're calling "injectSecret()" with a secret.
- First, its not really "injecting" it into anything, its creating it in etcd, right? So it seems like it should be called createSecret or storeSecret - inject made me think it was inserting into some other object.
- Second, why do we have injectSecret() as a new func instead of just inlining those 2 lines of code? Seems like an unnecessary split of logic.

And while I'm at it :-)  why do we have an injectionSet at all? Why isn't it just a Secret?  There's just the Secret in there anyway.  Do we really need something between the Binding and the Secret?

Comment (duglin) [inline pkg/controller/injector/k8s_binding_injector.go]: Net, it seems to me we could simplify this entire file (from here on) down to just two funcs: InjectBinding and UninjectBinding - and all bits of code could be inlined making it much easier to follow (and of course killing injectionSet).

Comment (duglin) [inline pkg/controller/injector/k8s_binding_injector.go]: @arschles if you don't want to do that reorg in this PR, while in here anyway, that's ok - I can follow-up with another PR if you want.

Comment (arschles) [inline pkg/controller/injector/k8s_binding_injector.go]: I :100: agree. For context, I held off on making refactors in this PR, but may as well clean this file up.

Changes incoming

Comment (arschles) [inline pkg/controller/injector/k8s_binding_injector.go]: @duglin changes made in https://github.com/kubernetes-incubator/service-catalog/pull/256/commits/c3ef033a96a5a06c34fba7d6276d15fb0d33900f to simplify this file. thanks for prompting me

File: pkg/controller/injector/k8s_binding_injector.go [hunk 3]
```diff
@@ -98,4 +64,18 @@ func makeSecret(binding *servicecatalog.Binding, cred *brokerapi.Credential) *v1
 			"password": []byte(cred.Password),
 		},
 	}
+	// TODO: don't hardcode this namespace. https://github.com/kubernetes-incubator/service-catalog/issues/162
```
Comment (pmorie) [inline pkg/controller/injector/k8s_binding_injector.go]: I think we can talk through enough of the basics of #162 shortly that we can remove the hardcoding here, and would prefer to do that before this goes in.  Game?

File: pkg/controller/injector/k8s_binding_injector.go [hunk 4]
```diff
@@ -59,35 +53,7 @@ func CreateK8sBindingInjector() (BindingInjector, error) {
...
-	return &injectionSet{
-		secret: secret,
-	}
-}
-
-func makeSecret(binding *servicecatalog.Binding, cred *brokerapi.Credential) *v1.Secret {
-	return &v1.Secret{
+	secret := &v1.Secret{
 		ObjectMeta: v1.ObjectMeta{
 			Name: binding.Name,
```
Comment (pmorie) [inline pkg/controller/injector/k8s_binding_injector.go]: If the binding spec has `SecretName` set, we should use that instead of the binding name.

Comment (duglin) [inline pkg/controller/injector/k8s_binding_injector.go]: Agreed, but I don't see a SecretName on BindingSpec yet.

**Comment by duglin** [comment]
@arschles thanks a ton. @MHBauer already LGTM'd it so I can't add another label, but for the record this LGTM and is much easier to follow now.

File: pkg/controller/injector/k8s_binding_injector.go (hunk 3 above)
Comment (arschles) [inline pkg/controller/injector/k8s_binding_injector.go]: @pmorie I'll do it now, but in another PR

Comment (arschles) [inline pkg/controller/injector/k8s_binding_injector.go]: See https://github.com/kubernetes-incubator/service-catalog/pull/262

File: pkg/controller/injector/k8s_binding_injector.go (hunk 4 above)
Comment (arschles) [inline pkg/controller/injector/k8s_binding_injector.go]: Let's do in another PR to avoid scope creep here. I'm happy to make the change if someone makes an issue for it.

Comment (arschles) [inline pkg/controller/injector/k8s_binding_injector.go]: Also, I believe it's `SecretRef` in the binding spec

Comment (duglin) [inline pkg/controller/injector/k8s_binding_injector.go]: I don't think we'd pick up the name from SecretRef would we?  While I guess we could, I would think that if we're going to allow the user to pick the secret's name then it would be on the BindingSpec first, right?  Meaning, that's the source of truth and the name on/in the SecretRef is just a slave to it.  Right?

Comment (duglin) [inline pkg/controller/injector/k8s_binding_injector.go]: Actually, I think we'd have to grab it from the BindingSpec since during this create() we shouldn't have a valid SecretRef yet.

Comment (arschles) [inline pkg/controller/injector/k8s_binding_injector.go]: `SecretRef` is in the [binding spec](https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/apis/servicecatalog/types.go#L261) and points to the name of the secret to create with the binding creds

Comment (duglin) [inline pkg/controller/injector/k8s_binding_injector.go]: ok, education question... yea I know SecretRef points to the secret, but I was viewing this property more like a pointer, meaning its not set until we create the secret and then its set to whatever Kube uses for reference pointers.  Are you saying that SecretRef is a user-defined value?  Is this the normal pattern for Kube?

Comment (pmorie) [inline pkg/controller/injector/k8s_binding_injector.go]: SecretRef is the declaration of the name of the secret to create from the result of a binding.  I think it should default to the name of the binding if not specified.

Comment (duglin) [inline pkg/controller/injector/k8s_binding_injector.go]: ok, I think I finally understand why this simple thing was throwing me. The use of the word "Ref" in there implied something like a pointer (defined by the system) and therefore its something that is set after the secret is created.  What I think you guys are saying is that SecretRef isn't really system defined - its actually user defined but if not specified we'll overwrite it with the binding.Name.

If that's right... I then have a UX question... is this normally how we ask users to provide us the name of the resource they want us to create?  For example, I would have normally expected to see binding.SecretName as the field a user would set.  Coming at this fresh, a name like "SecretRef' would imply we're asking the user to give us the reference to an existing object, not the name (or ref) of a yet to be created object. But if this is a normal Kube thing then ok.  Just wondering.

Comment (pmorie) [inline pkg/controller/injector/k8s_binding_injector.go]: `SecretRef` might be a poorly-chosen name.  IIRC this is a name we came up with on the last day of the first F2F meeting.  Perhaps `SecretName` is better?

Naming of the field aside, we need to be able to know the name of the objects that will be created based on the Binding spec, so that when the Binding includes the SIP spec, that the content of the SIP spec can address those objects by name.  Does that make sense?

Comment (duglin) [inline pkg/controller/injector/k8s_binding_injector.go]: yep - I get the semantics of what we're doing, it was just the way it was manifesting itself in the objects that felt a little off - and yes, SecretName would be better IMO.

Comment (arschles) [inline pkg/controller/injector/k8s_binding_injector.go]: @duglin can you create an issue for changing to `SecretName`?
